src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,657 @@
+     * @throws InvalidAlgorithmParameterException
+     *         if the information contained within the {@code derivationSpec} is
+     *         invalid or if the combination of {@code alg} and the
+     *         {@code derivationSpec} results in something invalid
+     * @throws NoSuchAlgorithmException
+     *         if {@code alg} is empty or invalid
+     * @throws NullPointerException
+     *         if {@code alg} or {@code derivationSpec} is null
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider
+     *         Selection</a>
+     *
+     */
+    public SecretKey deriveKey(String alg,
+                               AlgorithmParameterSpec derivationSpec)
+            throws InvalidAlgorithmParameterException,
+                   NoSuchAlgorithmException {
+        if (alg == null) {
+            throw new NullPointerException(
+                    "the algorithm for the SecretKey return value must not be"
+                    + " null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                    "the algorithm for the SecretKey return value must not be "
+                    + "empty");
+        }
+        Objects.requireNonNull(derivationSpec);
+        if (checkSpiNonNull(theOne)) {
+            return theOne.spi().engineDeriveKey(alg, derivationSpec);
+        } else {
+            return (SecretKey) chooseProvider(alg, derivationSpec);
+        }
+    }
+
+    /**
+     * Derives a key, returns raw data as a byte array.
+     *
+     * @param derivationSpec
+     *         the object describing the inputs to the derivation function
+     *
+     * @return the derived key in its raw bytes
+     *
+     * @throws InvalidAlgorithmParameterException
+     *         if the information contained within the {@code derivationSpec} is
+     *         invalid
+     * @throws UnsupportedOperationException
+     *         if the derived keying material is not extractable
+     * @throws NullPointerException
+     *         if {@code derivationSpec} is null
+     *
+     * @see <a href="#DelayedProviderSelection">Delayed Provider
+     *         Selection</a>
+     *
+     */
+    public byte[] deriveData(AlgorithmParameterSpec derivationSpec)
+            throws InvalidAlgorithmParameterException {
+
+        Objects.requireNonNull(derivationSpec);
+        if (checkSpiNonNull(theOne)) {
+            return theOne.spi().engineDeriveData(derivationSpec);
+        } else {
+            try {
+                return (byte[]) chooseProvider(null, derivationSpec);
+            } catch (NoSuchAlgorithmException e) {
+                // this will never be thrown in the deriveData case
+                throw new AssertionError(e);
+            }
+        }
+    }
+
+    /**
+     * Use the firstSpi as the chosen KDFSpi and set the fields accordingly
+     */
+    private void useFirstSpi() {
+        if (checkSpiNonNull(theOne)) return;
+
+        synchronized (lock) {
```

Comment:
Sorry I missed this last time. The `if (checkSpiNonNull(theOne)` check also must be repeated inside the `synchronized` block.

Suppose 2 threads, one calling `deriveData` and one calling `getProviderName`, are running at the same time. Both start with `theOne` being null and reach the block. The `deriveData` thread gets the lock, chooses a provider, and assigns `theOne`. When it releases the lock the `getProviderName` thread enters this block and updated `theOne`. This should not happen.

Reply:
Have you thought about the test?